### PR TITLE
44 ma sam+ and more

### PR DESCRIPTION
**Pull request by oriKay**
ma sam plus
ma sam bug fixes
literal cnf bug fixes
tests
experiments
bugs with folds reading and solvers fixes.

### CODE REVIEW

File: sam_learning/learners/ma_sam_plus.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    unsafe_actions_preconditions_map: Dict[str, CompoundPrecondition]
+
+    def __init__(self, partial_domain: Domain,
+                 preconditions_fluent_map: Optional[Dict[str, List[str]]] = None,
+                 negative_precondition_policy: NegativePreconditionPolicy = NegativePreconditionPolicy.no_remove):
+        super().__init__(partial_domain, preconditions_fluent_map, negative_precondition_policy=negative_precondition_policy)
+        self.mapping = {}
+        self.unsafe_actions_preconditions_map = {}
+
+    def _extract_predicate_from_clause(self, clause_element: Tuple[str, str], mapping: BindingType) -> Predicate:
```
Comment (argaman-aloni) [inline sam_learning/learners/ma_sam_plus.py:32]: maybe rename

File: sam_learning/learners/ma_sam_plus.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        relevant_preconditions_str = {precondition.untyped_representation for action in lma_set for precondition
+                                      in action.preconditions if isinstance(precondition, Predicate)}
+
+        for fluent, fluent_cnf in self.literals_cnf.items():
+            effects = fluent_cnf.extract_macro_action_effects(lma_names, relevant_preconditions_str, param_grouping)
+            for effect_element in effects:
+                cnf_effects.append(self._extract_predicate_from_clause(effect_element, mapping))
+
+        # TODO find a neater way to remove duplicates
+        unique_representations = {}
```
Comment (argaman-aloni) [inline sam_learning/learners/ma_sam_plus.py:94]: Check this if need the True value in the list.

File: sam_learning/learners/ma_sam_plus.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+        new_precondition = CompoundPrecondition()
+        for action in action_group:
+            preconditions = action.preconditions if action.name not in self.unsafe_actions_preconditions_map \
+                else self.unsafe_actions_preconditions_map[action.name]
+
+            for _, precondition in preconditions:
+                if isinstance(precondition, Predicate):
+                    cnf_preconditions.append(
+                        self._extract_predicate_from_clause((action.name,
```
Comment (argaman-aloni) [inline sam_learning/learners/ma_sam_plus.py:121]: merge with "add precondition below".

File: sam_learning/learners/ma_sam_plus.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        super()._remove_unobserved_actions_from_partial_domain()
+        for action in self.partial_domain.actions.values():
+            self.logger.debug("Constructing safe action for %s", action.name)
+            action_preconditions = {precondition for precondition in
+                                    action.preconditions if isinstance(precondition, Predicate)}
+            if not self._is_action_safe(action, action_preconditions):
+                self.logger.warning("Action %s is not safe to execute!", action.name)
+                #TODO need to check if its copy or overriden next line
+                self.unsafe_actions_preconditions_map[action.name] = action.preconditions
+                action.preconditions = CompoundPrecondition()
```
Comment (argaman-aloni) [inline sam_learning/learners/ma_sam_plus.py:148]: you can remove the action completely

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 5]
```diff
@@ -3,6 +3,44 @@
...
+
+    return True
+
+
+def group_params_from_clause(clause) -> List[set]:
+    """
+    Processes a single clause, grouping parameters from different actions
+    by their index position in the match.
+    """
+    param_pattern = re.compile(r'\?\w+')
```
[the remark this answers is not in the document]
Comment (oriKay) [reply sam_learning/core/propositional_operations/literals_cnf.py:54]: Returns a match where the string contains any word characters (characters from a to Z, digits from 0-9, and the underscore _ character)

**Pull request merged**